moveit_kinematics/test/test_kinematics_plugin.cpp

```diff
@@ -0,0 +1,737 @@
+    kinematics_solver_->searchPositionIK(poses[0], seed, 0.1, consistency_limits, solution, error_code);
+    if (error_code.val != error_code.SUCCESS)
+    {
+      ++failures;
+      continue;
+    }
+
+    // on success: validate reached poses
+    std::vector<geometry_msgs::Pose> reached_poses;
+    kinematics_solver_->getPositionFK(tip_frames, solution, reached_poses);
+    EXPECT_NEAR_POSES(poses, reached_poses, IK_NEAR);
+
+    // validate closeness of solution pose to goal
+    auto diff = Eigen::Map<Eigen::ArrayXd>(solution.data(), solution.size()) -
+                Eigen::Map<Eigen::ArrayXd>(goal.data(), goal.size());
+    if (!diff.isZero(1.05 * NEAR_JOINT))
+    {
+      ++failures;
+      ROS_WARN_STREAM("jump in [" << i << "]: " << diff.transpose());
+    }
+
+    // update robot state to found pose
+    robot_state.setJointGroupPositions(jmg_, solution);
+    traj.addSuffixWayPoint(robot_state, 0.1);
+  }
+  EXPECT_LE(failures, (1.0 - EXPECTED_SUCCESS_RATE) * num_ik_tests_);
+
+  if (publish_trajectory)
+  {
+    ros::NodeHandle nh;
+    ros::AsyncSpinner spinner(1);
+    spinner.start();
+    ros::Publisher pub = nh.advertise<moveit_msgs::DisplayTrajectory>("display_random_walk", 1, true);
+    traj.getRobotTrajectoryMsg(msg.trajectory[0]);
+    pub.publish(msg);
+    ros::WallDuration(0.1).sleep();
+  }
+}
+
+static double parseDouble(XmlRpc::XmlRpcValue& v)
+{
+  if (v.getType() == XmlRpc::XmlRpcValue::TypeDouble)
+    return static_cast<double>(v);
+  else if (v.getType() == XmlRpc::XmlRpcValue::TypeInt)
+    return static_cast<int>(v);
+  else
+    return 0.0;
+}
+static void parseVector(XmlRpc::XmlRpcValue& vec, std::vector<double>& values, size_t num = 0)
+{
+  ASSERT_EQ(vec.getType(), XmlRpc::XmlRpcValue::TypeArray);
+  if (num != 0)
+    ASSERT_EQ(vec.size(), num);
+  values.reserve(vec.size());
+  values.clear();
+  for (size_t i = 0; i < vec.size(); ++i)
+    values.push_back(parseDouble(vec[i]));
+}
+static bool parseGoal(const std::string& name, XmlRpc::XmlRpcValue& value, Eigen::Isometry3d& goal)
+{
+  std::vector<double> vec;
+  if (name == "pose")
+  {
+    parseVector(value["orientation"], vec, 4);
+    goal = Eigen::Quaterniond(vec[3], vec[0], vec[1], vec[2]);  // w x y z
+    parseVector(value["position"], vec, 3);
+    goal.translation() = Eigen::Map<Eigen::Vector3d>(vec.data());
+    return true;
+  }
+  for (unsigned char axis = 0; axis < 3; ++axis)
+  {
+    char axis_char = 'x' + axis;
+    // position offset
+    if (name == (std::string("pos.") + axis_char))
+    {
+      goal.translation()[axis] += parseDouble(value);
+      return true;
+    }
+    // rotation offset
+    else if (name == (std::string("rot.") + axis_char))
+    {
+      goal *= Eigen::AngleAxisd(parseDouble(value), Eigen::Vector3d::Unit(axis));
+      return true;
+    }
+  }
+  return false;
+}
+
+TEST_F(KinematicsTest, unitIK)
+{
+  XmlRpc::XmlRpcValue tests;
+  if (!getParam("unit_tests", tests))
```

Comment:
Should this be wrapped in an assert?

Reply:
unit test definitions in the test launch are optional. If there are none defined, there is nothing to test.